orderbook/src/api/post_quote.rs

```diff
@@ -79,33 +112,210 @@ struct OrderQuote {
     buy_token_balance: BuyTokenDestination,
 }
 
+#[derive(Debug)]
+pub enum FeeAndQuoteError {
+    FeeError(FeeError),
+    OrderError(ValidationError),
+    // RateLimit, // TODO - use this.
+    Other(anyhow::Error),
+}
+
+impl FeeAndQuoteError {
+    pub fn convert_to_reply(self) -> (Json, StatusCode) {
+        match self {
+            FeeAndQuoteError::FeeError(err) => err.to_warp_reply(),
+            FeeAndQuoteError::OrderError(err) => err.to_warp_reply(),
+            // FeeAndQuoteError::RateLimit => (
+            //     super::error("RateLimit", "Too many order quotes"),
+            //     StatusCode::TOO_MANY_REQUESTS,
+            // ),
+            FeeAndQuoteError::Other(err) => {
+                tracing::error!(?err, "get_fee_and_quote error");
+                (super::internal_error(), StatusCode::INTERNAL_SERVER_ERROR)
+            }
+        }
+    }
+}
+
+impl OrderQuoteRequest {
+    fn build_quote_from_sell_response(
+        &self,
+        fee_response: SellResponse,
+    ) -> Result<OrderQuote, FeeAndQuoteError> {
+        match self.side {
+            OrderQuoteSide::Sell { sell_amount } => {
+                let sell_amount = match sell_amount {
+                    SellAmount::BeforeFee {
+                        value: sell_amount_before_fee,
+                    } => sell_amount_before_fee,
+                    SellAmount::AfterFee {
+                        value: sell_amount_after_fee,
+                    } => sell_amount_after_fee
+                        .checked_sub(fee_response.fee.amount)
+                        .ok_or(FeeAndQuoteError::FeeError(SellAmountDoesNotCoverFee))?
+                        .max(U256::one()),
+                };
+                Ok(OrderQuote {
+                    from: self.from,
+                    sell_token: self.sell_token,
+                    buy_token: self.buy_token,
+                    receiver: self.receiver,
+                    sell_amount,
+                    buy_amount: fee_response.buy_amount_after_fee,
+                    valid_to: self.valid_to,
+                    app_data: self.app_data,
+                    fee_amount: fee_response.fee.amount,
+                    kind: match self.side {
+                        OrderQuoteSide::Sell { .. } => OrderKind::Sell,
+                        OrderQuoteSide::Buy { .. } => OrderKind::Buy,
+                    },
+                    partially_fillable: self.partially_fillable,
+                    sell_token_balance: self.sell_token_balance,
+                    buy_token_balance: self.buy_token_balance,
+                })
+            }
+            OrderQuoteSide::Buy { .. } => Err(FeeAndQuoteError::Other(anyhow!(
+                "Can't build Quote for Buy Request with SellResponse"
+            ))),
```

Comment:
It feels a bit awkward to have multiple matches and require these unreachable branches. Maybe there is a way to refactor so that this isn't needed?

Reply:
Yea, I would hope so. I will try harder to think of one.

Reply:
Fixed and looks much better now. I had a feeling there was something off/silly about those matches.